### PR TITLE
Cache history feature

### PR DESCRIPTION
**Pull request by MaroderKo**

### CODE REVIEW

File: src/main/java/com/autosale/shop/controller/LoginController.java [hunk 1]
```diff
@@ -31,4 +31,10 @@ public ResponseEntity<JwtTokensDTO> refreshTokens(@RequestBody Map<String, Strin
     public boolean verifyToken(@RequestBody JwtTokensDTO tokens) {
         return authenticationService.isAuthenticated(tokens.getAccessToken());
     }
+
+    @PostMapping("/tokens/terminate/{id}")
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/controller/LoginController.java]: It is not about tokens. Should be another SessionController with another path

File: src/main/java/com/autosale/shop/listeners/CacheListener.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.event.EventListener;
+import org.springframework.data.redis.core.RedisKeyExpiredEvent;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class CacheListener {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/listeners/CacheListener.java]: SessingEventListener

File: src/main/java/com/autosale/shop/listeners/CacheListener.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+    private final LogsRepository logsRepository;
+
+    @EventListener
+    public void onSessionExpireEvent(RedisKeyExpiredEvent<?> event) {
+        logsRepository.log("Session of user " + new String(event.getId()).substring(1) + " is expired");
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/listeners/CacheListener.java]: 1. Use String format.
2. Explain your actions related to substring in Java doc

File: src/main/java/com/autosale/shop/listeners/CacheListener.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.context.event.EventListener;
+import org.springframework.data.redis.core.RedisKeyExpiredEvent;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class CacheListener {
+
+    private final LogsRepository logsRepository;
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/listeners/CacheListener.java]: Why LogsRepository integrated directly here? Each repository should has own service

File: src/main/java/com/autosale/shop/repository/LogsRepository.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.autosale.shop.repository;
+
+public interface LogsRepository {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/repository/LogsRepository.java]: 1. Not Logs but Log
2. I would name it as TraceRepository. or some TraceHistoryRepository
3. Should have own service
4. Suggest you to clean old traces once per day by cron job

File: src/main/java/com/autosale/shop/repository/LogsRepository.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+
+    void log(String message);
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/repository/LogsRepository.java]: method names in repositories should represent SQL operations. So you can name it as "create" or "insert"

File: src/main/java/com/autosale/shop/service/AuthenticationService.java [hunk 7]
```diff
@@ -12,5 +12,7 @@ public interface AuthenticationService {
 
     boolean isAuthenticated(String token);
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/AuthenticationService.java]: where is method is used?

File: src/main/java/com/autosale/shop/service/AuthenticationService.java [hunk 8]
```diff
@@ -12,5 +12,7 @@ public interface AuthenticationService {
...
 
-    Authentication getAuthentication(String token);
+    Authentication getAuthentication(User user);
+
+    void terminateSession(int id);
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/AuthenticationService.java]: not related to AuthenticationService. should be moved UserSessionService

File: src/main/java/com/autosale/shop/service/JwtTokenService.java [hunk 9]
```diff
@@ -2,12 +2,11 @@
...
 import com.autosale.shop.model.User;
-import io.jsonwebtoken.Claims;
 
 public interface JwtTokenService {
     JwtTokensDTO generateTokensPair(User user);
 
     JwtTokensDTO generateTokensFromRefreshToken(String refreshToken);
 
-    Claims getClaimsFromToken(String token);
+    User getUserFromToken(String token);
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/JwtTokenService.java]: parseUser(...)

File: src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java [hunk 10]
```diff
@@ -22,37 +22,39 @@ public class AuthenticationServiceImpl implements AuthenticationService {
 
     private final JwtTokenService jwtTokenService;
     private final UserService userService;
+    private final UserSessionService userSessionService;
 
     @Override
     public JwtTokensDTO generateTokensByUserCredentials(User user) {
-
         user = userService.getVerifiedUser(user.getUserName(), user.getPassword());
+        userSessionService.createSession(user);
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java:23]: session should be created in AuthenticationService. Session is not related to token generating

File: src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java [hunk 11]
```diff
@@ -22,37 +22,39 @@ public class AuthenticationServiceImpl implements AuthenticationService {
...
         return jwtTokenService.generateTokensPair(user);
     }
 
     @Override
     public JwtTokensDTO generateTokensByRefreshToken(String token) {
-        return jwtTokenService.generateTokensFromRefreshToken(token);
+        JwtTokensDTO jwtTokensDTO = jwtTokenService.generateTokensFromRefreshToken(token);
+        User userFromToken = jwtTokenService.getUserFromToken(token);
+        userSessionService.createSession(userFromToken);
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java:38]: same here

File: src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java [hunk 12]
```diff
@@ -22,37 +22,39 @@ public class AuthenticationServiceImpl implements AuthenticationService {
...
+        return jwtTokensDTO;
     }
 
     @Override
     public boolean isAuthenticated(String token) {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java]: Does this method still in using?

File: src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java [hunk 13]
```diff
@@ -22,37 +22,39 @@ public class AuthenticationServiceImpl implements AuthenticationService {
...
-        }
-        return claims.get("username", String.class).equals(user.getUserName()) && claims.get("role", String.class).equals(user.getRole().name()) && claims.getExpiration().after(new Date());
+        return userSessionService.getSession(jwtTokenService.getUserFromToken(token)) != null;
     }
 
     @Override
-    public Authentication getAuthentication(String token) {
-        Claims claims = jwtTokenService.getClaimsFromToken(token);
-        return new UsernamePasswordAuthenticationToken(claims.get("id", String.class), "", getAuthorities(claims));
+    public Authentication getAuthentication(User user) {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java]: remove this method. Do it in filter

File: src/main/java/com/autosale/shop/service/impl/UserSessionServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+import lombok.RequiredArgsConstructor;
+import org.springframework.cache.annotation.CacheEvict;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Service;
+
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+public class UserSessionServiceImpl implements com.autosale.shop.service.UserSessionService {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/service/impl/UserSessionServiceImpl.java]: use import  com.autosale.shop.service

**Review by YaroslavMovchan — COMMENTED** [review]
Structurize your application layers based on their responsibilities

File: src/main/java/com/autosale/shop/listeners/CacheListener.java (hunk 3 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/listeners/CacheListener.java]: Event returns if in binary array format and has at start char ':', substring removes this ':' from converted string with user id

File: src/main/java/com/autosale/shop/service/AuthenticationService.java (hunk 7 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/service/AuthenticationService.java]: Now it has special endpoint /login/tokens/verify that is used more for testing than for release functionality.

File: src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java (hunk 10 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java:23]: actually, methods that create tokens are the only way to log in for users, this code is in AuthenticationService implementation, what should I do then?

File: src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java (hunk 12 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java]: same answer as in it's interface

File: src/main/java/com/autosale/shop/service/impl/UserSessionServiceImpl.java (hunk 14 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/service/impl/UserSessionServiceImpl.java]: just a side effect after automatically extracting the interface from the class

File: src/main/java/com/autosale/shop/listeners/CacheListener.java (hunk 3 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/listeners/CacheListener.java]: substring dont have sens anymore, removed

File: src/main/java/com/autosale/shop/service/AuthenticationService.java (hunk 7 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/service/AuthenticationService.java]: removed

File: src/main/java/com/autosale/shop/listeners/SessionEventListener.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+@Component
+@RequiredArgsConstructor
+public class SessionEventListener {
+
+    private final TraceHistoryService traceHistoryService;
+
+    @EventListener
+    public void onSessionExpireEvent(RedisKeyExpiredEvent<?> event) {
+        traceHistoryService.log(String.format("Session of user %s is expired", new String(event.getId())));
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/listeners/SessionEventListener.java:17]: String format takes obj as a second argument. Don't need to do new String(event.getId()))

File: src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+@Repository
+@RequiredArgsConstructor
+public class UserSessionRepositoryImpl implements UserSessionRepository {
+
+    private final LettuceConnectionFactory lettuceConnectionFactory;
+
+    private final GenericJackson2JsonRedisSerializer serializer;
+
+    @Value("${jwt.access-token-lifetime}")
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java]: init in constructor

File: src/main/java/com/autosale/shop/config/RedisObjectMappingConfiguration.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.data.redis.serializer.GenericJackson2JsonRedisSerializer;
+
+@Configuration
+public class RedisObjectMappingConfiguration {
+
+    @Bean
+    GenericJackson2JsonRedisSerializer userSessionRedisSerializer() {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/config/RedisObjectMappingConfiguration.java]: Do we really need this? Probably we can use objectMapper?

File: src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+    @Override
+    public UserSession read(String key) {
+        try (RedisConnection connection = lettuceConnectionFactory.getConnection()) {
+            byte[] bytes = connection.stringCommands().get(key.getBytes());
+            if (bytes == null) {
+                return null;
+            }
+            Object result = serializer.deserialize(bytes);
+            if (result instanceof UserSession) {
+                return (UserSession) result;
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java]: Don't like this casting. Probably we can use ObjectMapper to mapping bytes to exact object

File: src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+
+    @Override
+    public void create(String key, UserSession session) {
+        try (RedisConnection connection = lettuceConnectionFactory.getConnection()) {
+            connection.stringCommands().setEx(key.getBytes(), sessionTimeToLive * 60L, serializer.serialize(session));
+        }
+    }
+
+    @Override
+    public UserSession read(String key) {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java]: use Optional

File: src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+                return (UserSession) result;
+            }
+            return null;
+        }
+    }
+
+    @Override
+    public boolean delete(String key) {
+        try (RedisConnection connection = lettuceConnectionFactory.getConnection()) {
+            return connection.commands().expireAt(key.getBytes(), 1);
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java:50]: don't we have simple command just for deleting? Or you want to listen to this event?

File: src/main/java/com/autosale/shop/security/JwtAuthenticationFilter.java [hunk 21]
```diff
@@ -21,16 +27,19 @@ public class JwtAuthenticationFilter extends OncePerRequestFilter {
...
     @Override
     protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
 
         Optional<String> tokenOptional = extractToken(request);
         if (tokenOptional.isPresent()) {
             String token = tokenOptional.get();
-            if (authenticationService.isAuthenticated(token)) {
-                Authentication authentication = authenticationService.getAuthentication(token);
+            UserSession session = userSessionService.getSession(jwtTokenService.parseUser(token));
+            if (session != null) {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/security/JwtAuthenticationFilter.java]: optional

File: src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java (hunk 10 above)
Comment (YaroslavMovchan) [reply src/main/java/com/autosale/shop/service/impl/AuthenticationServiceImpl.java:23]: ok, leave it here

File: src/main/java/com/autosale/shop/listeners/SessionEventListener.java (hunk 15 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/listeners/SessionEventListener.java:17]: Yeah, but event.getId returns an array, that is not an object, so anyway we need to cast, somehow, array of bytes to object and I think that casting it directly to string is the best way

File: src/main/java/com/autosale/shop/config/RedisObjectMappingConfiguration.java (hunk 17 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/config/RedisObjectMappingConfiguration.java]: Changed to ObjectMapper but to ignore JsonProcessingException (which is not really so important https://github.com/FasterXML/jackson-databind/issues/723) I marked methods with SneakyThrows annotations

Comment (YaroslavMovchan) [reply src/main/java/com/autosale/shop/config/RedisObjectMappingConfiguration.java:11]: But it still throws :) So, I recommend handling an exception

File: src/main/java/com/autosale/shop/listeners/SessionEventListener.java (hunk 15 above)
Comment (YaroslavMovchan) [reply src/main/java/com/autosale/shop/listeners/SessionEventListener.java:17]: Aha, I see. It returns an array of bytes. Then good

File: src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java (hunk 20 above)
Comment (MaroderKo) [reply src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java:50]: I want to listen to this event even if it's administrator session termination, just for logging purposes. If I log all user's session termination why should I make an exception for this special situation?

File: src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+            connection.stringCommands().setEx(key.getBytes(), sessionTimeToLive * 60L, serializer.writeValueAsBytes(session));
+        }
+    }
+
+    @Override
+    @SneakyThrows
+    public Optional<UserSession> read(String key) {
+        try (RedisConnection connection = lettuceConnectionFactory.getConnection()) {
+            byte[] bytes = connection.stringCommands().get(key.getBytes());
+            if (bytes == null) {
```
Comment (YaroslavMovchan) [inline src/main/java/com/autosale/shop/repository/impl/UserSessionRepositoryImpl.java]: simplify using Optional.ofnullable

**Review by YaroslavMovchan — APPROVED** [review]

**Pull request merged**